llvm/test/Transforms/LoopVectorize/trip-count-expansion-may-introduce-ub.ll

```diff
@@ -527,6 +528,76 @@ exit:
   ret i64 %p
 }
 
+define i64 @multi_exit_4_exit_count_with_udiv_by_frozen_value_in_latch(ptr %dst, i64 %N) {
```

Comment:
I think SCEV is actually computing the trip count wrong here.  It currently produces `((42 /u %fr.N) umin (0 smax %N2))`... but I think it should actually be `((0 smax %N2) umin_seq (42 /u %fr.N))`.  And that should be a well-defined if %N2 is zero, even if we would divide by zero without the umin_seq.  (I don't think anyone was thinking about divide-by-zero when https://reviews.llvm.org/D116766 was originally proposed, but it seems like a natural extension.)

Under those semantics, we shouldn't need to modify VPlan to examine the exits.  SCEVExpander itself should enable SafeUDivMode when it evaluates operands of umin_seq after the first.

Reply:
(The relevant code on the SCEV side is ScalarEvolution::getSequentialMinMaxExpr; it "optimizes" umin_seq to umin.)

Reply:
So the proposal here is to change the semantics of umin_seq from a select to a conditional branch, so it not only prevents poison propagation but also UB propagation?

Does that really cover all cases? What if we have one exit with a udiv exit count, but also a prior abnormal exit? Maybe not relevant to vectorization, but thinking more generally here.

Reply:
I'm not sure there's really a useful distinction between an abnormal exit inside the loop, vs. an abnormal exit in the loop preheader.  Either way, if you want to hoist the trip count past such an operation, you need to take additional measures beyond what SCEV normally provides.  You need safe-udiv for all udiv operations, and if you want to use the value for anything, you need to freeze it.

For simple vectorization, we should be able to assume the loop exits normally, and we want to use that information to avoid unnecessary freeze operations.  So we need to distinguish which freeze/safe-div operations are necessary.  The way we currently figure out which operands we need to freeze is via umin_seq; I think it makes sense to figure out which operands need safe-div the same way.

Reply:
Updated the PR to use umin_seq in the cases as suggested I think; the main additional change is to forbid simplifying `A umin_seq B` -> `A umin B`, if `B` may trigger UB (at the moment just SCEVUDivExpr?). This could be split off separately + SCEV printing test. Just want to make sure this is what you had in mind before doing so.

(There are some redundant freezes of the UDiv result which we could get rid of)

Looking at this again, I am not sure if the divisor being poison in `@multi_exit_4_exit_count_with_udiv_by_value_in_latch` actually needs freezing; as the expanded expression is used for a branch, if `%N` is poison, the branch on poison will also trigger UB, just slightly later. Granted at the SCEV &  expander level, we don't really know if that will be the case. `@multi_exit_4_exit_count_with_udiv_by_value_in_latch_different_bound` uses 2 different bounds for the exits, so here the freeze of the divisor is definitely needed

Reply:
Yes, this is what I was thinking, roughly.

> I am not sure if the divisor being poison in @multi_exit_4_exit_count_with_udiv_by_value_in_latch actually needs freezing

We could probably refine this in SCEVExpander if we end up generating a bunch of freeze in inconvenient places (basically, make a map of SCEVUnknowns which are used without freezing them, and then don't freeze if we see them elsewhere).